### PR TITLE
Fix contract import.

### PR DESCRIPTION
**Pull request by yspringfield**
Fixed: #7984

### CODE REVIEW

File: libs/utils/src/lib/spreadsheet/import.ts [hunk 1]
```diff
@@ -242,13 +245,13 @@ export function getStaticList<S extends Scope>(scope: S, value: string, separato
...
+  territories: (territories, separator, required:boolean) => getStaticList('territories', territories, separator, 'Territories', required, 'world'),
+  medias: (medias, separator, required:boolean) => getStaticList('medias', medias, separator, 'Medias', required, 'all'),
 }
 
-export function getGroupedList(value: string, groupScope: 'territories', separator: string): FromStatic<'territories'>;
-export function getGroupedList(value: string, groupScope: 'medias', separator: string): FromStatic<'medias'>;
-export function getGroupedList<GS extends GroupScope>(value: string, groupScope: GS, separator: string) {
+export function getGroupedList(value: string, groupScope: 'territories', separator: string, options?:GroupedListOptions): FromStatic<'territories'>;
+export function getGroupedList(value: string, groupScope: 'medias', separator: string, options?:GroupedListOptions): FromStatic<'medias'>;
+export function getGroupedList<GS extends GroupScope>(value: string, groupScope: GS, separator: string, options={mandatory:true}) {
```
Comment (RemcoSimonides) [inline libs/utils/src/lib/spreadsheet/import.ts:254]: Fix space here too
`options?: GroupListOptions `

`options = { mandatory: true }`

File: libs/movie/src/lib/movie/+state/movie.service.ts [hunk 2]
```diff
@@ -90,6 +96,18 @@ export class MovieService extends CollectionService<MovieState> {
     });
   }
 
+  getImportTitle({ name, userOrgId, blockframesAdmin, path}: ImportTitleOptions) {
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/+state/movie.service.ts:99]: There is not reason to put that in the service as it's only used in the import system. Put it back in the import utils.

File: libs/utils/src/lib/spreadsheet/import.ts [hunk 3]
```diff
@@ -14,6 +14,9 @@ export interface SheetTab {
   headers: any[];
   rows: any[][];
 }
+interface GroupedListOptions{
+  mandatory:boolean;
```
[the remark this answers is not in the document]
Comment (GrandSchtroumpf) [reply libs/utils/src/lib/spreadsheet/import.ts:18]: Why `mandatory` while everywhere else it's "required" ?

File: libs/import/src/lib/utils.ts [hunk 4]
```diff
@@ -83,17 +83,16 @@ export async function getTitleId(
...
-      return ref.where('title.international', '==', name);
-    } else {
-      return ref
-        .where('title.international', '==', name)
-        .where('orgIds', 'array-contains', userOrgId);
-    }
-  });
-  const result = titles.length === 1 ? titles[0].id : '';
+  const options = { path: 'title.international', blockframesAdmin, userOrgId, name };
+  let titles = await titleService.getImportTitle(options);
```
Comment (GrandSchtroumpf) [inline libs/import/src/lib/utils.ts:87]: stop using `let` for something else than a string or a number, it's very difficult to understand.

File: libs/import/src/lib/utils.ts [hunk 5]
```diff
@@ -74,28 +74,43 @@ export async function getOrgId(
...
-  cache[name] = result;
-  return result;
+}
+
+
+export async function getTitleId(
+  name: string,
+  titleService: MovieService,
+  cache: Record<string, Movie>,
+  memo: (name: string, value: Movie) => string,
```
Comment (GrandSchtroumpf) [inline libs/import/src/lib/utils.ts:94]: Why do you pass the memo function ? Define it in the scope of this function

File: libs/import/src/lib/view-extracted-elements/contract/utils.ts [hunk 6]
```diff
@@ -111,21 +112,37 @@ export async function formatContract(
 ) {
   // Cache to avoid  querying db every time
   const orgNameCache: Record<string, string> = {};
-  const titleNameCache: Record<string, string> = {};
+  const titleNameCache: Record<string, Movie> = {};
   const userCache: Record<string, User> = {};
   const contractCache: Record<string, Mandate | Sale> = {};
+  const memo = (key: string, value: Movie) => {
```
Comment (GrandSchtroumpf) [inline libs/import/src/lib/view-extracted-elements/contract/utils.ts:118]: Move that to the getTitleId function as a closure

File: libs/movie/src/lib/movie/+state/movie.service.ts [hunk 7]
```diff
@@ -27,6 +27,12 @@ export const fromInternalRef = (internalRef: string): QueryFn => (ref) =>
 type MovieWithAnalytics = Movie & { analytics: MovieAnalytics };
 
 interface MovieState extends EntityState<Movie, string>, ActiveState<string> { }
+export interface ImportTitleOptions {
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/+state/movie.service.ts:30]: you don't need that anymore

**Review by GrandSchtroumpf — APPROVED** [review]

**Pull request merged**